### PR TITLE
Generic numeric debugging (#19317) (#19317)

### PR DESCRIPTION
**Pull request by metascroy**
Summary: Pull Request resolved: https://github.com/pytorch/executorch/pull/19317

Pulled By:
metascroy

metascroy

Differential Revision: D103956056

### CODE REVIEW

File: devtools/intermediate_output_tap/_tap_pass.py [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    output_node = graph.output_node()
+    existing_outputs = list(output_node.args[0])
+
+    # Snapshot before we start mutating the graph.
+    candidate_nodes = [n for n in graph.nodes if not _is_tap_node(n)]
+
+    specs: list[TapSpec] = []
+    new_tap_nodes: list[fx.Node] = []
+
+    for node in candidate_nodes:
```
Comment (digantdesai) [inline devtools/intermediate_output_tap/_tap_pass.py:119]: what happens if the delegate have to fuse some candidate nodes? Try lowering conv2d-->batch_norm, where XNNPACK doesn't support standalone batch_norm IIRC.
Also how is it better than forced-single-op-partitions?

Comment (metascroy) [reply devtools/intermediate_output_tap/_tap_pass.py:119]: This is primarily to help RL in a numeric debugging investigation. I may clean it up to make it a generic utility if they find it useful, but will likely go through design review to get input from others if I do that.

To answer your question on fused candidates: this is tested with CoreML's quantized linear pattern [dequantize -> linear]. In that case, we tap the intermediate output after linear node, which is actually a quantized linear in both eager (from QDQ pattern) and CoreML (from its internal fusion). In the [conv2d-->batch_norm] case, I'd have to check. Tapping batch norm should mean we want the output of batch_norm, i.e., the intermediate output after batchnorm, which can be the result of a fused [conv2d-->batch_norm] op. If we did forced single-op partitions on con2d and batch_norm separately, we wouldn't get the fusion.

> how is it better than forced-single-op-partitions

One reason is single-op partitions destroy fusion that backends do. Here we tap intermediates, so as long as we tap the final intermediate after a fusion pattern we should be good. 

A second reason is this approach does keep the same big delegate blob, just with extra outputs. In CoreML's case that means it will still be routed to ANE, whereas if you break up into single op partitions, it will very likely reroute to CPU b/c they are small.

**Review by YIWENX14 — APPROVED** [review]
LGTM.

File: devtools/intermediate_output_tap/_tap_pass.py (hunk 1 above)
Comment (metascroy) [reply devtools/intermediate_output_tap/_tap_pass.py:119]: I think RL has found this useful, so I may actually try landing. I added an E2E test for XNNPACK with quantized Conv2D + Batch norm

**Review by JacobSzwejbka — APPROVED** [review]
